doc/adr007-flattening.md

```diff
@@ -0,0 +1,255 @@
+  2["(2) val a = N"] --> 1
+  3["(3) val b = B::A1::a"] --> 6
+  4["(4) val c = C::A1::a"] --> 8
+  5["(5) pure val B::A1::N = 1"]
+  6["(6) pure val B::A1::a = B::A1::N"] --> 5
+  7["(7) pure val C::A1::N = 0"]
+  8["(8) pure val C::A1::a = C::A1::N"] --> 7
+```
+
+##### Override expressions
+
+In an instance statement like `import A(N=e).*`, `e` might depend on expressions defined in the current module. In this case, we need to ensure that the dependencies of `e` are copied to the new module created for the instance, otherwise `pure val myModule::A::N = e` won't resolve. To accomplish this, we use the actual `Flattener` (described below) to inspect only the expressions in the overrides and find any definitions they might depend on.
+
+### Flattener
+
+The flattener adds missing definitions to a module by recursively fetching the lookup table and copying the referenced definition, then looking into that definition for any other names and doing the same.
+
+When adding a missing definition, if that definitions carries a namespace, we actually need to do a "deep" procedure to add that name to not only the definition name, but also to every other name appearing in it's body. This includes adding the namespace to lambda parameters and nested definitions (even though they cannot be referred), so flattening doesn't create name conflicts with them.
+
+This deep namespace-adding procedure is already done in the current implementation of flattening (see [`addNamespaceToDef`](https://github.com/informalsystems/quint/blob/main/quint/src/flattening.ts#L276C4-L276C4)), with the only difference being that we don't want to generate new ids in this case (we only create new ids in the Instance Flattener).
+
+Although the Instance Flattener requires imports and instances to be flattened, we can't flatten definitions that come from instances before running the instance flattener, since that would copy definitions that would potentially be replaced by new definitions in new modules created by the instance flattener. Therefore, we need to run the flattener twice: first not copying definitions that come from instances (checking the `importedFrom` field once again), and then, after the instance flattener, flattening anything that remains to be flattened.
+
+### Overall flattening process
+
+The flattening process is run completely for each module before proceeding to flatten the next module. The modules are assumed to be topologically sorted. The need for running this by module is that we do remove the import statements after we resolve all dependencies, but we can only ensure that all dependencies were flattened if the depending modules were completely flattened before.
+
+The whole flattening process is composed by the following steps:
+1. Run the Flattener, ignoring definitions that come from instances
+2. Remove `import` and `export` statements
+3. Run the Instance Flattener
+4. Resolve names for the module and its dependencies, obtaining a new lookup table
```

Comment:
What keeps us from producing the updated lookup table as a by-product of flattening?

Reply:
My cognitive abilities, mainly :laughing:. I think the same note I made for item (6) is valid here, with some added difficulty: step (3) adds new ids, so we would have to update all lookup table pointers involving those ids.

My reasoning is: we need to have proper names anyway since apalache has it's own name resolution thing (afaiu from your's and Igor's response when I asked about that need). And, if we have the names correct, than all we have to do to get our lookup table correct is run the name resolver.

In another direction, we could just get our lookup table to be right, and then ensure we have proper (unique) names by ensuring each id in the lookup table has a different name. I tried doing that, but is was much easier for me to reason about modules and names (which you can print and visualize) than with lookup table pointers.

Reply:
I added a note to this point